src/reader/warming.rs

```diff
@@ -0,0 +1,383 @@
+use std::{
+    collections::HashSet,
+    iter::FromIterator,
+    sync::{Arc, Mutex, Weak},
+    thread::JoinHandle,
+    time::Duration,
+};
+
+use crate::{Executor, Searcher, SegmentId, TantivyError};
+
+pub const GC_INTERVAL: Duration = Duration::from_secs(1);
+
+/// `Warmer` can be used to maintain segment-level state e.g. caches.
+///
+/// They must be registered with the [IndexReaderBuilder].
+pub trait Warmer: Sync + Send {
+    /// Perform any warming work using the provided [Searcher].
+    fn warm(&self, searcher: &Searcher) -> crate::Result<()>;
+
+    /// Discard internal state for any [SegmentId] not provided.
+    fn garbage_collect(&self, live_segment_ids: &HashSet<SegmentId>);
+}
+
+/// Warming-related state with interior mutability.
+#[derive(Clone)]
+pub(crate) struct WarmingState(Arc<Mutex<WarmingStateInner>>);
+
+impl WarmingState {
+    pub fn new(num_warming_threads: usize, warmers: Vec<Weak<dyn Warmer>>) -> crate::Result<Self> {
+        Ok(Self(Arc::new(Mutex::new(WarmingStateInner {
+            num_warming_threads,
+            warmers,
+            searcher_generations: Vec::new(),
+            gc_thread: None,
+        }))))
+    }
+
+    /// Start tracking a new generation of searchers, and [Warmer::warm] it if there are active warmers.
+    ///
+    /// A background GC thread for [Warmer::garbage_collect] calls is uniquely created if there are active warmers.
+    pub fn new_searcher_generation(
+        &self,
+        liveness_token: Weak<()>,
+        searcher: &Searcher,
+    ) -> crate::Result<()> {
+        self.0
+            .lock()
+            .unwrap()
+            .new_searcher_generation(liveness_token, searcher, &self.0)
+    }
+
+    #[cfg(test)]
+    fn gc_maybe(&self) -> bool {
+        self.0.lock().unwrap().gc_maybe()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SearcherGeneration {
+    liveness_token: Weak<()>,
+    segment_ids: Vec<SegmentId>,
+}
+
+impl SearcherGeneration {
+    fn is_live(&self) -> bool {
+        self.liveness_token.strong_count() > 0
+    }
+}
+
+#[derive(Default)]
+struct WarmingStateInner {
+    num_warming_threads: usize,
+    warmers: Vec<Weak<dyn Warmer>>,
+    searcher_generations: Vec<SearcherGeneration>,
```

Comment:
Just came across the https://docs.rs/census/latest/census/ crate, which tantivy already uses...